core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java

```diff
@@ -14,133 +14,384 @@
-    } catch (URISyntaxException e) {
-      throw new RuntimeException(e);
+  UfsJournal(URI location, UnderFileSystem ufs) {
+    mLocation = URIUtils.appendPathOrDie(location, VERSION);
+    mUfs = ufs;
+
+    mLogDir = URIUtils.appendPathOrDie(mLocation, LOG_DIRNAME);
+    mCheckpointDir = URIUtils.appendPathOrDie(mLocation, CHECKPOINT_DIRNAME);
+    mTmpDir = URIUtils.appendPathOrDie(mLocation, TMP_DIRNAME);
+  }
+
+  @Override
+  public URI getLocation() {
+    return mLocation;
+  }
+
+  @Override
+  public JournalReader getReader(JournalReaderCreateOptions options) {
+    return new UfsJournalReader(this, options);
+  }
+
+  @Override
+  public JournalWriter getWriter(JournalWriterCreateOptions options) throws IOException {
+    if (options.isPrimary()) {
+      return new UfsJournalLogWriter(this, options);
+    } else {
+      return new UfsJournalCheckpointWriter(this, options);
+    }
+  }
+
+  @Override
+  public boolean isFormatted() throws IOException {
+    UnderFileSystem ufs = UnderFileSystem.Factory.get(mLocation.toString());
+    UnderFileStatus[] files = ufs.listStatus(mLocation.toString());
+    if (files == null) {
+      return false;
     }
+    // Search for the format file.
+    String formatFilePrefix = Configuration.get(PropertyKey.MASTER_FORMAT_FILE_PREFIX);
+    for (UnderFileStatus file : files) {
+      if (file.getName().startsWith(formatFilePrefix)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public void format() throws IOException {
+    URI location = getLocation();
+    LOG.info("Formatting {}", location);
+    if (mUfs.isDirectory(location.toString())) {
+      for (UnderFileStatus p : mUfs.listStatus(location.toString())) {
+        URI childPath = URIUtils.appendPathOrDie(location, p.getName());
+        boolean failedToDelete;
+        if (p.isDirectory()) {
+          failedToDelete = !mUfs
+              .deleteDirectory(childPath.toString(), DeleteOptions.defaults().setRecursive(true));
+        } else {
+          failedToDelete = !mUfs.deleteFile(childPath.toString());
+        }
+        if (failedToDelete) {
+          throw new IOException(String.format("Failed to delete %s", childPath));
+        }
+      }
+    } else if (!mUfs.mkdirs(location.toString())) {
+      throw new IOException(String.format("Failed to create %s", location));
+    }
+
+    // Create a breadcrumb that indicates that the journal folder has been formatted.
+    UnderFileSystemUtils.touch(URIUtils.appendPathOrDie(location,
+        Configuration.get(PropertyKey.MASTER_FORMAT_FILE_PREFIX) + System.currentTimeMillis())
+        .toString());
   }
 
   /**
-   * @return the location of the journal checkpoint
+   * @return the log directory location
    */
-  protected URI getCheckpoint() {
-    try {
-      return URIUtils.appendPath(mLocation, CHECKPOINT_FILENAME);
-    } catch (URISyntaxException e) {
-      throw new RuntimeException(e);
-    }
+  URI getLogDir() {
```

Comment:
I don't think we use package-private methods anywhere else in our code base so I would consider adding `protected` here to be consistent